### ISSUE TITLE
Entry not found in cache

### ISSUE THREAD

**Issue by chrmarti** [Microsoft/vscode-azure-account#53]
From @jdneo in https://github.com/Microsoft/vscode-azure-account/issues/45#issuecomment-380321789:

Hi @chrmarti I found out a way to reproduce `Entry not found in cache` error. Just trigger the `Azure: Open bash in clould shell` twice. 

More details can be found in the gif below:

![report](https://user-images.githubusercontent.com/6193897/38596047-efb293b2-3d81-11e8-8de5-4d263d2314f8.gif)

**Comment by chrmarti** [comment]
@jdneo I cannot reproduce this myself currently. Is it still a problem and would it be fixed by the PR AzureAD/azure-activedirectory-library-for-nodejs#188?

**Comment by jdneo** [comment]
@chrmarti 

I can still repro this on my side, the steps are:
- Make sure the Azure Account is signed out at first
- Trigger `Open bash in Cloud Shell`
- Sign in
- Wait for the Cloud Shell connected
- Dispose the Cloud Shell terminal
- Trigger `Open bash in Cloud Shell` again

**Comment by chrmarti** [comment]
That reproduces it and the PR you linked seems to fix it. Now we just need that fix in adal-node.

**Comment by derek1ee** [comment]
I was able to repro this, simply by leaving VS Code open for ~an hour without interacting with it.

**Comment by su-jie** [comment]
We are experiencing the same issue. Do you guys know if it has been fixed?

**Comment by chrmarti** [comment]
No, https://github.com/AzureAD/azure-activedirectory-library-for-nodejs/issues/197 should be reopened from what I see.

**Comment by fiveisprime** [comment]
Any idea on the right person/team to contact about getting this fixed?

**Comment by chrmarti** [comment]
@navyasric closed the upstream issue (AzureAD/azure-activedirectory-library-for-nodejs#197), maybe we can reach out to her.

**Comment by gi-joe-moto** [comment]
Still encountering this issue...

**Comment by ms-guizha** [comment]
any one can help for this issue ?

**Comment by chrmarti** [comment]
Upstream is unresponsive. We could look into shipping a patched version, but we also want to look into moving to MSAL (replacing ADAL that shows the problem).

**Comment by e-compton** [comment]
If anyone is still encountering this issue, I found that deleting the slsTokenCache worked for me:
```
rm ~/.azure/slsTokenCache.json
```

https://github.com/serverless/serverless-azure-functions/issues/412#issuecomment-591511201

**Comment by scale-tone** [comment]
What an ever-green issue. ~100 mentions! 
Let me contribute to it as well.

I'm occasionally seeing this error when for some reason the token cache (maintained by **adal-node** library) contains **two** valid tokens with the same **resource** value:
![image](https://user-images.githubusercontent.com/5447190/141700739-cdb40b2f-a8dc-4a63-9152-77f48d880c66.png)

This makes `_loadSingleEntryFromCache()` method produce an error as shown on the screenshot, which is then converted into an `Entry not found in cache` failure:
![image](https://user-images.githubusercontent.com/5447190/141700969-9e0a374e-a318-4cd6-af97-6a4f7f66b2b1.png)

I don't know which chain of events leads to a token being duplicated in the cache. I also don't know why **adal-node** treats this condition as a failure.

**adal-node** v0.2.3
**ms-vscode.azure-account-0.9.11**

**Comment by scale-tone** [comment]
> I don't know which chain of events leads to a token being duplicated in the cache

No, wait, I know! 
It's a concurrency issue. Token cache (or some code that uses it) doesn't tolerate parallel async calls, that use the same kind of token. 
In my case the solution was to always make all network calls sequentially. 
But there're always chances for unrelated calls to interfere with each other like that, so this is definitely a bug on the token cache side.
